internal/ethapi/api.go

```diff
@@ -1024,6 +1064,80 @@ func (s *BlockChainAPI) Call(ctx context.Context, args TransactionArgs, blockNrO
 	return result.Return(), result.Err
 }
 
+// BatchCallConfig is the config object to be passed to eth_batchCall.
+type BatchCallConfig struct {
+	Block          rpc.BlockNumberOrHash
+	StateOverrides *StateOverride
+	Calls          []BatchCallArgs
+}
+
+// BatchCallArgs is the object specifying each call within eth_batchCall. It
+// extends TransactionArgs with the list of block metadata overrides.
+type BatchCallArgs struct {
+	TransactionArgs
+	BlockOverrides *BlockOverrides
+}
+
+// CallResult is the result of one call.
+type CallResult struct {
+	Return hexutil.Bytes
+	Error  error
+}
```

Comment:
I really like that this is finally being implemented in the core protocol since it makes it very easy to do a number of things.

It'd be really great if this also includes gas consumed by the particular call. It'd help to set more appropriate gas limits when someone needs to sign the next transactions while the initial state-changing transactions are pending/not broadcasted.

Reply:
If it's useful we can easily return the gas used here. But can you please elaborate on your use-case? I didn't quite understand.

Reply:
An example of simple use-case: `erc20 approve` + `defi interaction`. `eth_estimateGas` done before the `erc20 approve` getting confirmed would usually revert. Hence the normal flow of UX is to send `approve` tx and wait for it to confirm then do the `defi interaction` tx. This is well known to be not good UX because of waiting for approve to be confirmed. Using huge input gas limit is not the good solution because wallets show max eth fees (gas price * gas limit) which looks costly and some users might not have enough eth.

`eth_batchCall` can be used instead of `eth_estimateGas` where the `erc20 approve` tx is mentioned as first call and then the `defi interaction` as second call and it's `gasUsed` can be used to very accurately estimate gas even before prervious transaction is confirmed. The improved UX for this becomes: click on button in dapp once, hit confirm on metamask/wallet twice, check back in like few mins if both txs are confirmed. 

Another interesting one is: If user had to do a lot of steps like approve + deposit + stake + what not, a dapp could use `eth_batchCall` to simulate the UI state after a user interaction and create list of txs to submit and get them signed at once and accurately estimate the gas limit (similar to github PR reviews where we can add lot of comments while scrolling at our convenience and it gets submited all at once).

TLDR including gasUsed basically creates the functionality of batched version of `eth_estimateGas`. I hope the usecase makes sense.

Reply:
Ok now I understand the use-case, thanks for extensive description. I think adding `gasUsed` to the result is not a good solution. If you look, logic of `eth_estimateGas` is more complicated than simply doing a `eth_call` and reporting the gasUsed. This AFAIK is because the gaslimit provided to the tx can change the flow of the tx itself (`GAS` opcode).

But the use-case is valid IMO and warrants a `eth_batchEstimateGas` or something of the sort.

Reply:
@s1na I had a implementation for `batchEstimateGas` way way back ago https://github.com/ethereum/go-ethereum/pull/21268

But I think your approach is much better, would be appreciate you can also take over this with the same design(e.g. state overrides, block overrides, etc)

Reply:
Yeah, it makes sense. So if `gas` is not specified in a call object it'd assume a large value, in order to ensure complex state-changing calls follow a successful execution path if there exists any, wouldn't users need to use `eth_batchEstimateGas` (or something like that) for setting the `gas` field in the calls prior to using `eth_batchCall`?

Reply:
> An example of simple use-case: `erc20 approve` + `defi interaction`.
> 
> Problem: `eth_estimateGas` done before the `erc20 approve` getting confirmed would usually revert. Hence the normal flow of UX is to send `approve` tx and wait for it to confirm then do the `defi interaction` tx. This is well known to be not good UX because of increased user waiting times. Using huge input gas limit is not the good solution because wallets show max eth fees (gas price * gas limit) which looks costly and some users might not have enough eth.
> 
> Solution: if `eth_batchCall` includes `gasUsed`, then it can be used instead of `eth_estimateGas` where the `erc20 approve` tx is mentioned as first call and then the `defi interaction` as second call and it's `gasUsed` can be used to very accurately estimate gas even before prervious transaction is confirmed. The improved UX for this becomes: click on button in dapp once, hit confirm on metamask/wallet twice, check back in like few mins if both txs are confirmed. Similarly, if the user had to do a lot of steps like approve + deposit + stake + what not, a dapp could use `eth_batchCall` to simulate the UI state after a user interaction and create list of txs to submit and get them signed at once and accurately estimate the gas limit (similar to github PR reviews where we can add lot of comments while scrolling at our convenience and it gets submited all at once). This saves a lot of user's waiting time, and hence has the potential to improve the UX considerably.
> 
> TLDR including gasUsed basically enables estimating gas on a state updated after a series of calls. I hope the usecase makes sense.
> 
> Edit: I just came across a project (created by Uniswap engineer) that exposes an endpoint for batch estimateGas using mainnet fork ([link](https://github.com/marktoda/gaspipe)), use-case mentioned in their README beginning is exactly what I am trying to explain above.

gas used through a wrapper contract is not accurate with Multicall due to EIP-2929, so should be avoided FYI (this is why Uniswap made this endpoint i think)